src/main/scala/akka/stream/contrib/ZipInputStreamSource.scala

```diff
@@ -0,0 +1,206 @@
+  in:        () => ZipInputStream,
+  chunkSize: Int                  = DefaultChunkSize
+) extends GraphStageWithMaterializedValue[SourceShape[(ZipEntryData, ByteString)], Future[Long]] {
+
+  val matValue = Promise[Long]()
+
+  override val shape =
+    SourceShape(Outlet[(ZipEntryData, ByteString)]("zipInputStreamSource.out"))
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+
+    val InputBuffer(initialBuffer, maxBuffer) =
+      inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16))
+
+    val logic = new GraphStageLogic(shape) {
+      import shape._
+
+      private var is: ZipInputStream = _
+      private var readBytesTotal: Long = 0L
+      private var buffer: Vector[(ZipEntryData, ByteString)] = Vector.empty
+      private var eof: Boolean = false
+      private var currentEntry: Option[ZipEntry] = None
+      private var currentStreams: Seq[ZipInputStream] = Seq()
+
+      override def preStart(): Unit = {
+        super.preStart()
+        try {
+          is = in()
+          currentStreams = List(is)
+          fillBuffer(initialBuffer)
+        } catch {
+          case ex: Exception =>
+            matValue.failure(ex)
+            failStage(ex)
+        }
+      }
+
+      setHandler(out, new OutHandler {
+
+        override def onPull(): Unit = {
+          fillBuffer(maxBuffer)
+          buffer match {
+            case Seq() =>
+              finalize()
+            case head +: Seq() =>
+              push(out, head)
+              finalize()
+            case head +: tail =>
+              push(out, head)
+              buffer = tail
+          }
+          def finalize() = try {
+            is.close()
+          } finally {
+            matValue.success(readBytesTotal)
+            complete(out)
+          }
+        }
+
+        override def onDownstreamFinish(): Unit =
+          try {
+            is.close()
+          } finally {
+            matValue.success(readBytesTotal)
+            super.onDownstreamFinish()
+          }
+      }) // end of handler
+
+      @tailrec private def nextEntry(
+        streams: Seq[ZipInputStream]
+      ): (Option[ZipEntry], Seq[ZipInputStream]) = streams match {
+        case Seq() => (None, streams)
+        case (z :: zs) =>
+          val entry = Option(z.getNextEntry)
+          entry match {
+            case None =>
+              nextEntry(zs)
+            case Some(e) if isZipFile(e) =>
+              nextEntry(new ZipInputStream(z) +: streams)
+            case Some(e) if e.isDirectory =>
+              nextEntry(streams)
+            case _ =>
+              (entry, streams)
+          }
+      }
+
+      private def isZipFile(e: ZipEntry) = e.getName.toLowerCase.endsWith(".zip")
+
+      /** BLOCKING I/O READ */
+      private def readChunk() = {
+        def read(arr: Array[Byte]) = currentStreams.headOption.flatMap { stream =>
+          val readBytes = stream.read(arr)
```

Comment:
To be honest I'm not familiar with the `scala.concurrent.blocking` idiom. According to [this discussion](http://stackoverflow.com/questions/29068064/scala-concurrent-blocking-what-does-it-actually-do) the real benefits strongly depend on the `ExecutionContext` involved. I've tried to be a good citizen of Akka :) and followed the @ktoso 's ["Zen of Akka"](http://es.slideshare.net/ktoso/zen-of-akka) by assigning the `IODispatcher` thread pool by default in the attributes. I've run a rough search in the Akka codebase but I couldn't find examples of use of this idiom. So, I'm not very comfortable with this change. Could anyone else throw some light on the matter?.